File: Longest-Substring-Without-Repeating-Characters.js
/**
 * @param {string} s
 * @return {number}
 */
// .charCodeAt(0) gives the ascii value of the character
// Ref: https://www.youtube.com/watch?v=-zSxTJkcdAo
var lengthOfLongestSubstring = function (s) {
  let l = 0,
    r = 0,
    n = s.length,
    mx = 0,
    hash = new Array(256).fill(-1);
    while (r < n) {
        if (hash[s[r].charCodeAt(0)] != -1) {
            if (hash[s[r].charCodeAt(0)] >= l) {
                l = hash[s[r].charCodeAt(0)] + 1;
            }
        }
        const len = r - l + 1;
        mx = Math.max(len, mx);
        hash[s[r].charCodeAt(0)] = r++;
    }
    return mx;
};

File: Longest-Substring-Without-Repeating-Characters.py
1class Solution:
2    def lengthOfLongestSubstring(self, s: str) -> int:
3        n, res = len(s), 0
4        hm = [-1 for _ in range(1, 257)]
5        l = r = 0
6        while r < n:
7            if hm[ord(s[r])] != -1:
8                if hm[ord(s[r])] >= l:
9                    l = hm[ord(s[r])] + 1
10            ln = r - l + 1
11            res = max(res, ln)
12            hm[ord(s[r])] = r
13            r += 1
14        return res
15            